File: java/src/main/java/lambda/ServiceThree.java
package lambda;
import saaf.Inspector;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.s3.model.ObjectMetadata;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Map.Entry;

import java.io.*;

import java.util.*;

/**
 * uwt.lambda_test::handleRequest
 * 
 * @author wlloyd
 * @author david perez
 */
public class ServiceThree implements RequestHandler<Request, HashMap<String, Object>> {


        public HashMap<String, Object> handleRequest(Request request, Context context) {
        Inspector inspector = new Inspector();
        inspector.inspectAll();

        HashMap<String, String[]> filterByMap = new HashMap<String, String[]>();
        filterByMap.put("Region", new String[] { "Australia and Oceania" });
        filterByMap.put("Item Type", new String[] { "Office Supplies" });
        filterByMap.put("Sales Channel", new String[] { "Office Supplies" });
        filterByMap.put("Order Priority", new String[] { "Offline" });
        filterByMap.put("Country", new String[] { "Fiji" });

        HashMap<String, String[]> aggregateByMap = new HashMap<String, String[]>();
        aggregateByMap.put("max", new String[] { "Units Sold" });
        aggregateByMap.put("min", new String[] { "Units Sold" });
        aggregateByMap.put("avg", new String[] { "Order Processing Time", "Gross Margin", "Units Sold" });
        aggregateByMap.put("sum", new String[] { "Units Sold", "Total Revenue", "Total Profit" });

        String bucketname = request.getBucketName();
        String mytable = request.getTableName();
        String myTable = request.getTableName();
        //int size = hashSize(filterByMap);
        //String[] whereStrings = createWhereString(filterByMap);
        //String[] queryStrings = createAggFunctionStrings(aggregateByMap, myTable, whereStrings, size);
        //String fullQuery = Union_Queries(queryStrings);


        String fullQuery = constructQueryString(aggregateByMap, filterByMap, myTable);
        try {
            Properties properties = new Properties();
            properties.load(new FileInputStream("db.properties"));
            String username = System.getenv("username"); 
            String password = System.getenv("password"); 
            // String databaseName = System.getenv("databaseName");
            String url = System.getenv("url");

            Connection con = DriverManager.getConnection(url, username, password);
            PreparedStatement ps = con.prepareStatement(fullQuery);
            ResultSet rs = ps.executeQuery();
            StringBuilder queryResults = convertRStoCSV(rs);
            byte[] bytes = queryResults.toString().getBytes();
            InputStream is = new ByteArrayInputStream(bytes);
            ObjectMetadata meta = new ObjectMetadata();
            meta.setContentLength(bytes.length);
            meta.setContentType("text/plain");

            AmazonS3 s3Client = AmazonS3ClientBuilder.standard().build();
            s3Client.putObject(bucketname, "QueryResults.csv", is, meta);
            stressTest(10, mytable);
            con.close();
        } catch (Exception e) {
            System.out.println("Got an exception working with MySQL! " + e.getMessage());
        }

        inspector.addAttribute("value", "Finished with querying database");
        inspector.inspectAllDeltas();
        return inspector.finish();
    }

    public String constructQueryString(HashMap<String, String[]> aggregateValues, HashMap<String, String[]> filterValues, String tablename) {
        String aggr = "";
        for (Entry<String, String[]> e : aggregateValues.entrySet()) {
            for (String val: e.getValue()) {
                aggr += e.getKey().toUpperCase();
                aggr += "(`";
                String temp = val;
                aggr += temp;
                aggr += "`), ";

            }
        }
        String fil = "";
        for (Entry<String, String[]> e : filterValues.entrySet()) {
            for (String val : e.getValue()) {
                fil += "SELECT ";
                fil += aggr;
                fil += "'WHERE ";
                String temp_key = e.getKey();
                fil += temp_key;
                fil += "=";
                String temp_val = val;
                fil += temp_val;
                fil += "' AS `Filtered By` FROM ";
                fil += tablename;
                fil += " WHERE `";
                fil += temp_key;
                fil += "`='";
                fil += temp_val;
                fil += "' UNION ";

                
            }
        }
        String strippedWord = " UNION "; 
        StringBuilder strb=new StringBuilder(fil);    
        int index=strb.lastIndexOf(strippedWord); 
        strb.replace(index,strippedWord.length()+index,";");   
 
        return strb.toString();
    }
    private StringBuilder convertRStoCSV(ResultSet rs) throws Exception {
        StringBuilder sb = new StringBuilder();
        try {
            int numCols = rs.getMetaData().getColumnCount();
            while (rs.next()) {
                for (int i = 1; i <= numCols - 1; i++) {
                    sb.append(String.format(String.valueOf(rs.getString(i))) + ", ");

                }
                sb.append(String.format(String.valueOf(rs.getString(numCols))) + "\n");
            }

        } catch (SQLException e) {
            System.out.println("Got an exception working with convertRStoCSV! " + e.getMessage());
        }
        return sb;

    }

    public static void stressTest(int iterations, String mytable) {
        try {
            Connection con = DriverManager.getConnection(System.getenv("url"), System.getenv("username"),
                    System.getenv("password"));
            
            for (int i = 0; i < iterations; i++) {
                String testQuery = "SELECT * from " + mytable + ";";
                PreparedStatement ps = con.prepareStatement(testQuery);
                ResultSet rs = ps.executeQuery();
                int numCols = rs.getMetaData().getColumnCount();
                while (rs.next()) {
                    for (int j = 1; j <= numCols; j++) {
                        rs.getString(j);
                    }
                }
            }
            con.close();
        } catch (SQLException e) {
            System.out.println("Got an exception working with stressTest! " + e.getMessage());
        }
    }
}

File: nodejs/src/serviceThree/serviceThree.js
const RequestHandler = require('../requestHandler');
const Inspector = require('../Inspector');
const DbColumns = require('../common/dbColumns');
const DbUtils = require('../common/dbUtils');
const CsvUtils = require('../common/csvUtils');

/**
 * @typedef ServiceThreeRequest
 * @property {String} bucketname
 * @property {String} key
 * @property {String} dbName
 * @property {String} dbEndpoint
 * @property {String} tableName
 * @property {Number} stressTestLoops
 */

class ServiceThree extends RequestHandler {

    /**
     * @type {function(String):void}
     */
    _logger;

    /**
     * @type {S3}
     */
    _s3;

    /**
     * @type {MySQLClient}
     */
    _mysql;

    /**
     * @type {Object}
     */
    _environment;

    /**
     * @param {function(String):void} logger
     * @param {S3} s3
     * @param {MySQLClient} mysql
     * @param {Object} environment
     */
    constructor(logger, s3, mysql, environment) {
        super();
        this._logger = logger;
        this._s3 = s3;
        this._mysql = mysql;
        this._environment = environment;
    }

    /**
     *
     * @param {ServiceThreeRequest} request
     * @param context
     * @return {Promise<Object>}
     */
    async handleRequest(request, context) {
        let inspector = new Inspector();
        inspector.inspectAll();

        try {
            let query = this._constructQuery(request.tableName);

            /**
             * @type {Connection}
             */
            let connection = await this._mysql.createConnection({
                host: request.dbEndpoint,
                database: request.dbName,
                user: this._environment.username,
                password: this._environment.password,
                charset: 'utf8mb4',
                multipleStatements: false
            });

            let result = await connection.query(query);

            let buffer = Buffer.from(CsvUtils.fromRecords(result.map(row => DbUtils.fromObject(row))));

            await this._s3.putObject({
                Bucket: request.bucketname,
                Key: request.key,
                Body: buffer
            }).promise();

            await this._stressTest(connection, request);

            await connection.end();
        } catch (e) {
            this._log("I/O problem occurred when trying to finish the task: " + (typeof e === 'string' ? e : (e.message + "\n" + e.stack)));
            inspector.addAttribute("value", "Finished with querying database");
            inspector.inspectAllDeltas();
            return inspector.finish();
        }

        this._log("Successfully finished processing request");

        inspector.addAttribute("value", "Finished with querying database");
        inspector.inspectAllDeltas();
        return inspector.finish();
    }

    /**
     * @param {Connection} connection
     * @param {ServiceThreeRequest} request
     * @return {Promise<void>}
     * @private
     */
    async _stressTest(connection, request) {
        for (let i = 0; i < request.stressTestLoops; i++) {
            this._log(String(i));
            await connection.query("SELECT * FROM " + request.tableName + ";");
        }
    }

    /**
     * @param {String} tableName
     * @return {String}
     * @private
     */
    _constructQuery(tableName) {
        let filter = {};
        filter[DbColumns.REGION.identifier] = ["Australia and Oceania"];
        filter[DbColumns.ITEM_TYPE.identifier] = ["Office Supplies"];
        filter[DbColumns.SALES_CHANNEL.identifier] = ["Offline"];
        filter[DbColumns.PRIORITY.identifier] = ["Critical"];
        filter[DbColumns.COUNTRY.identifier] = ["Fiji"];

        return this._constructUnionSql(tableName, filter, {
            "max": [DbColumns.UNITS_SOLD.identifier],
            "min": [DbColumns.UNITS_SOLD.identifier],
            "avg": [DbColumns.ORDER_PROCESSING_TIME.identifier, DbColumns.GROSS_MARGIN.identifier, DbColumns.UNITS_SOLD.identifier],
            "sum": [DbColumns.UNITS_SOLD.identifier, DbColumns.REVENUE.identifier, DbColumns.PROFIT.identifier]
        });
    }

    /**
     * @param {String} table
     * @param {{String: Array<String>}} filter
     * @param {{String:Array<String>}} aggregate
     * @returns {String}
     * @private
     */
    _constructUnionSql(table, filter, aggregate) {
        let projection = Object.keys(aggregate)
            .map(fn => aggregate[fn].map(col => fn.toUpperCase() + "(" + col + ")").join(", "))
            .join(", ");
        return Object.keys(filter)
            .map(col => filter[col].map(value => this._constructSql(table, projection, col, value)))
            .join(" UNION ") + ";";
    }

    _constructSql(table, projection, filterColumn, filterValue) {
        return "SELECT " + projection + ", " +
            "'WHERE " + filterColumn.replace(/^`|`$/g, '') + "=" + filterValue + "' AS `Filtered By`"
            + " FROM " + table + " WHERE " + filterColumn + "= '" + filterValue + "'";
    }

    /**
     * @param {String} message
     * @private
     */
    _log(message) {
        this._logger.call(null, message);
    }

}

module.exports = ServiceThree;